### PR TITLE
Customshapes

### PR DESCRIPTION
**Pull request by vayyappaneni**
- Added the customStages class to pass the GH argument (list of stages) through cmdline
- Added the testrun argument --stop-timeout to gracefully shutdown the simulated users

### CODE REVIEW

File: src/grasshopper/lib/configuration/gh_configuration.py [hunk 1]
```diff
@@ -177,6 +177,16 @@ class ConfigurationConstants:
             "default": 120.0,
             "typecast": typecast_float,
         },
+        "stop_timeout": {
```
Comment (alteryx-sezell) [inline src/grasshopper/lib/configuration/gh_configuration.py:180]: nit: this name is confusing, it's a graceful shutdown time, correct? can you try a name along those lines?

File: src/grasshopper/lib/util/shapes.py [hunk 2]
```diff
@@ -196,3 +197,52 @@ def __init__(self, *args, **kwargs):
...
         ]
         super().__init__(*args, **kwargs)
+
+class Customstages(Default):  # noqa E501
+    """
+    Stolen from this set of examples as part of the locust.io documentation.
+    https://github.com/locustio/locust/blob/master/examples/custom_shape/stages.py
+
+    Keyword arguments:
+        stages -- A list of dicts, each representing a stage with the following keys:
```
Comment (alteryx-sezell) [inline src/grasshopper/lib/util/shapes.py:207]: this class takes a keyword argument `stages` that is a json string with this structure. it's important to note that.

File: src/grasshopper/lib/util/shapes.py [hunk 3]
```diff
@@ -196,3 +197,52 @@ def __init__(self, *args, **kwargs):
...
+            stage.
+            users -- Total user count
+            spawn_rate -- Number of users to start/stop per second
+            stop -- A boolean that can stop that test at a specific stage
+        stop_at_end -- Can be set to stop once all stages have run.
+
+    Most likely, you'd want to extend this class and only define a new stages attr.
+    """
+
+    stages = [{"duration": 68, "users": 4, "spawn_rate": 0.5},
```
Comment (alteryx-sezell) [inline src/grasshopper/lib/util/shapes.py:218]: you can feel free to have the default for this class be much simpler if you want, the default on the original was only there to provide an example

File: src/grasshopper/lib/util/shapes.py [hunk 4]
```diff
@@ -196,3 +197,52 @@ def __init__(self, *args, **kwargs):
...
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        try:
+            stages = json.loads(kwargs.get("stages"))
+            self.stages = stages
+        except TypeError:
+            pass
+        self._configured_runtime = self.stages[-1].get("duration", 0)
+
+    def tick(self):
```
Comment (alteryx-sezell) [inline src/grasshopper/lib/util/shapes.py:236]: you don't need this method. you are inheriting from `Stages` which has the exact same implementation. that is, in fact, the primary reason to inherit from that class.

File: src/grasshopper/lib/util/shapes.py [hunk 5]
```diff
@@ -196,3 +197,52 @@ def __init__(self, *args, **kwargs):
             },
         ]
         super().__init__(*args, **kwargs)
+
+class Customstages(Default):  # noqa E501
+    """
+    Stolen from this set of examples as part of the locust.io documentation.
```
Comment (alteryx-sezell) [inline src/grasshopper/lib/util/shapes.py:203]: nit: you don't really need this comment, it is relevant to the `Stages` class

File: tests/unit/conftest.py [hunk 6]
```diff
@@ -46,6 +46,7 @@ def current_global_defaults():
         "shape": "Default",
         "users": 1.0,
         "runtime": 120.0,
+        "stop_timeout": 300.0,
```
Comment (alteryx-sezell) [inline tests/unit/conftest.py:49]: 5 minutes for a graceful shutdown of users seems very long? at least for a default value.

i can't seem to find where this value is being used for anything? where is the code for that?

File: src/grasshopper/lib/configuration/gh_configuration.py (hunk 1 above)
Comment (vayyappaneni) [reply src/grasshopper/lib/configuration/gh_configuration.py]: Removed the argument from this PR, need more time to validate it.

File: src/grasshopper/lib/util/shapes.py (hunk 5 above)
Comment (vayyappaneni) [reply src/grasshopper/lib/util/shapes.py]: Removed

File: tests/unit/conftest.py (hunk 6 above)
Comment (vayyappaneni) [reply tests/unit/conftest.py]: I did remove the test arg from this PR.

File: src/grasshopper/lib/util/shapes.py (hunk 2 above)
Comment (vayyappaneni) [reply src/grasshopper/lib/util/shapes.py]: Updated

File: src/grasshopper/lib/util/shapes.py (hunk 3 above)
Comment (vayyappaneni) [reply src/grasshopper/lib/util/shapes.py:206]: Updated

**Review by alteryx-sezell — APPROVED** [review]

**Pull request merged**